### ISSUE TITLE
In `dplyr::filter()`, specifying a logical condition with `~ .` notation fails inside `purrr::map2()` but works otherwise

### ISSUE THREAD

**Issue by emmansh** [tidyverse/dplyr#6161]
Given this data frame
```r
my_mtcars <- tibble::rownames_to_column(mtcars, "cars")
```
those are the rows I want to keep:
```r
cars_to_keep <- c("Merc 240D", "Fiat X1-9", "Ferrari Dino")
```
and using `dplyr::filter()`, both of the following syntaxes work:
```r
# option a: specify `cars` colname explicitly
dplyr::filter(my_mtcars, cars %in% cars_to_keep)
# option b: use a selection helper and `~ .` notation
dplyr::filter(my_mtcars, across(contains("cars"), ~ . %in% cars_to_keep))
```

---
By contrast, when having a tibble with _list-columns_ (which suggests using `purrr:map2()` for filtering) only **option a** works
```r
# the data
higher_level_tibble <- 
  tibble::tibble(my_data         = list(my_mtcars),
                 the_cars_i_want = list(cars_to_keep))

# this works:
higher_level_tibble %>%
  dplyr::mutate(my_filtered_data = purrr::map2(.x = my_data, .y = the_cars_i_want, .f = ~dplyr::filter(.x, cars %in% .y)))

## # A tibble: 1 x 3
##   my_data        the_cars_i_want my_filtered_data
##  <list>         <list>          <list>          
## 1 <df [32 x 12]> <chr [3]>       <df [3 x 12]>  

# this doesn't:
higher_level_tibble %>%
  dplyr::mutate(my_filtered_data = purrr::map2(.x = my_data, .y = the_cars_i_want, .f = ~ dplyr::filter(.x, across(contains("cars"), ~ . %in% .y))))

## Error: Problem with `mutate()` column `my_filtered_data`.
## i `my_filtered_data = map2(...)`.
## x Problem with `filter()` input `..1`.
## i Input `..1` is `across(contains("cars"), ~. %in% .y)`.
## x the ... list contains fewer than 2 elements
```

---
The only way to make this work inside `dplyr::mutate()` & `purrr::map2()` is to replace the `~ .` notation inside `dplyr::filter()` with an anonymous function:
```r
higher_level_tibble %>%
  dplyr::mutate(my_filtered_data = purrr::map2(.x = my_data, .y = the_cars_i_want, .f = ~ dplyr::filter(.x, across(starts_with("cars"), function(x) x %in% .y))))

## # A tibble: 1 x 3
##   my_data        the_cars_i_want my_filtered_data
##   <list>         <list>          <list>          
## 1 <df [32 x 12]> <chr [3]>       <df [3 x 12]> 
```
Bottom line
---
How come `~ .` doesn't work but `function(x)` does when using `dplyr::filter()` inside `dplyr::mutate()` & `purrr::map2()`? This is particularly confusing as `~ .` does work outside of `purrr::map2()` context.

N.B.
See [this StackOverflow post](https://stackoverflow.com/questions/70772004/how-to-combine-filteracrossstarts-withfoo-logical-condition-with-mu/70773282#70773282) about the same issue.

**Comment by emmansh** [comment]
Reprex
---

``` r
library(dplyr, warn.conflicts = FALSE)
library(purrr)
library(tibble)

my_mtcars <- tibble::rownames_to_column(mtcars, "cars")

cars_to_keep <- c("Merc 240D", "Fiat X1-9", "Ferrari Dino")

# option a: specify `cars` colname explicitly
dplyr::filter(my_mtcars, cars %in% cars_to_keep)
#>           cars  mpg cyl  disp  hp drat    wt qsec vs am gear carb
#> 1    Merc 240D 24.4   4 146.7  62 3.69 3.190 20.0  1  0    4    2
#> 2    Fiat X1-9 27.3   4  79.0  66 4.08 1.935 18.9  1  1    4    1
#> 3 Ferrari Dino 19.7   6 145.0 175 3.62 2.770 15.5  0  1    5    6
# option b: use a selection helper and `~ .` notation
dplyr::filter(my_mtcars, across(contains("cars"), ~ . %in% cars_to_keep))
#>           cars  mpg cyl  disp  hp drat    wt qsec vs am gear carb
#> 1    Merc 240D 24.4   4 146.7  62 3.69 3.190 20.0  1  0    4    2
#> 2    Fiat X1-9 27.3   4  79.0  66 4.08 1.935 18.9  1  1    4    1
#> 3 Ferrari Dino 19.7   6 145.0 175 3.62 2.770 15.5  0  1    5    6

higher_level_tibble <- 
  tibble::tibble(my_data         = list(my_mtcars),
                 the_cars_i_want = list(cars_to_keep))

# works
higher_level_tibble %>%
  dplyr::mutate(my_filtered_data = purrr::map2(.x = my_data, .y = the_cars_i_want, .f = ~dplyr::filter(.x, cars %in% .y)))
#> # A tibble: 1 x 3
#>   my_data        the_cars_i_want my_filtered_data
#>   <list>         <list>          <list>          
#> 1 <df [32 x 12]> <chr [3]>       <df [3 x 12]>

# fails
higher_level_tibble %>%
  dplyr::mutate(my_filtered_data = purrr::map2(.x = my_data, .y = the_cars_i_want, .f = ~ dplyr::filter(.x, across(contains("cars"), ~ . %in% .y))))
#> Error: Problem with `mutate()` column `my_filtered_data`.
#> i `my_filtered_data = purrr::map2(...)`.
#> x Problem with `filter()` input `..1`.
#> i Input `..1` is `across(contains("cars"), ~. %in% .y)`.
#> x the ... list contains fewer than 2 elements

# works
higher_level_tibble %>%
  dplyr::mutate(my_filtered_data = purrr::map2(.x = my_data, .y = the_cars_i_want, .f = ~ dplyr::filter(.x, across(starts_with("cars"), function(x) x %in% .y))))
#> # A tibble: 1 x 3
#>   my_data        the_cars_i_want my_filtered_data
#>   <list>         <list>          <list>          
#> 1 <df [32 x 12]> <chr [3]>       <df [3 x 12]>


sessioninfo::session_info()
#> - Session info ---------------------------------------------------------------
#>  setting  value
#>  version  R version 4.1.2 (2021-11-01)
#>  os       Windows 10 x64 (build 19044)
#>  system   x86_64, mingw32
#>  ui       RTerm
#>  language (EN)
#>  collate  English_United States.1252
#>  ctype    English_United States.1252
#>  tz       Asia/Jerusalem
#>  date     2022-01-19
#>  pandoc   2.14.0.3 @ C:/Program Files/RStudio/bin/pandoc/ (via rmarkdown)
#> 
#> - Packages -------------------------------------------------------------------
#>  package     * version    date (UTC) lib source
#>  assertthat    0.2.1      2019-03-21 [1] CRAN (R 4.0.3)
#>  backports     1.4.1      2021-12-13 [1] CRAN (R 4.1.2)
#>  cli           3.1.0      2021-10-27 [1] CRAN (R 4.1.1)
#>  crayon        1.4.2      2021-10-29 [1] CRAN (R 4.1.1)
#>  DBI           1.1.2      2021-12-20 [1] CRAN (R 4.1.2)
#>  digest        0.6.29     2021-12-01 [1] CRAN (R 4.1.2)
#>  dplyr       * 1.0.7      2021-06-18 [1] CRAN (R 4.0.5)
#>  ellipsis      0.3.2      2021-04-29 [1] CRAN (R 4.0.5)
#>  evaluate      0.14       2019-05-28 [1] CRAN (R 4.0.3)
#>  fansi         1.0.2      2022-01-14 [1] CRAN (R 4.1.2)
#>  fastmap       1.1.0      2021-01-25 [1] CRAN (R 4.1.1)
#>  fs            1.5.2      2021-12-08 [1] CRAN (R 4.1.2)
#>  generics      0.1.1      2021-10-25 [1] CRAN (R 4.1.1)
#>  glue          1.6.0      2021-12-17 [1] CRAN (R 4.1.2)
#>  highr         0.9        2021-04-16 [1] CRAN (R 4.1.1)
#>  htmltools     0.5.2      2021-08-25 [1] CRAN (R 4.1.1)
#>  knitr         1.37       2021-12-16 [1] CRAN (R 4.1.2)
#>  lifecycle     1.0.1      2021-09-24 [1] CRAN (R 4.1.1)
#>  magrittr      2.0.1      2020-11-17 [1] CRAN (R 4.0.3)
#>  pillar        1.6.4      2021-10-18 [1] CRAN (R 4.1.1)
#>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.0.3)
#>  purrr       * 0.3.4      2020-04-17 [1] CRAN (R 4.0.3)
#>  R.cache       0.15.0     2021-04-30 [1] CRAN (R 4.1.1)
#>  R.methodsS3   1.8.1      2020-08-26 [1] CRAN (R 4.1.1)
#>  R.oo          1.24.0     2020-08-26 [1] CRAN (R 4.1.1)
#>  R.utils       2.11.0     2021-09-26 [1] CRAN (R 4.1.1)
#>  R6            2.5.1      2021-08-19 [1] CRAN (R 4.1.1)
#>  reprex        2.0.1.9000 2021-11-03 [1] Github (tidyverse/reprex@b96611f)
#>  rlang         0.4.12     2021-10-18 [1] CRAN (R 4.1.1)
#>  rmarkdown     2.11       2021-09-14 [1] CRAN (R 4.1.1)
#>  rstudioapi    0.13       2020-11-12 [1] CRAN (R 4.0.3)
#>  sessioninfo   1.2.2      2021-12-06 [1] CRAN (R 4.1.2)
#>  stringi       1.7.6      2021-11-29 [1] CRAN (R 4.1.2)
#>  stringr       1.4.0      2019-02-10 [1] CRAN (R 4.0.5)
#>  styler        1.6.2      2021-09-23 [1] CRAN (R 4.1.1)
#>  tibble      * 3.1.6      2021-11-07 [1] CRAN (R 4.1.2)
#>  tidyselect    1.1.1      2021-04-30 [1] CRAN (R 4.1.1)
#>  utf8          1.2.2      2021-07-24 [1] CRAN (R 4.1.1)
#>  vctrs         0.3.8      2021-04-29 [1] CRAN (R 4.0.5)
#>  withr         2.4.3      2021-11-30 [1] CRAN (R 4.1.2)
#>  xfun          0.29       2021-12-14 [1] CRAN (R 4.1.2)
#>  yaml          2.2.1      2020-02-01 [1] CRAN (R 4.0.3)
#> 
#>  [1] C:/Program Files/R/my_packages
#>  [2] C:/Program Files/R/R-4.1.2/library
#> 
#> ------------------------------------------------------------------------------
```

<sup>Created on 2022-01-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by hadley** [comment]
Somewhat more minimal reprex:

``` r
library(dplyr, warn.conflicts = FALSE)
library(purrr)

df <- tibble(
  df = list(tibble::rownames_to_column(mtcars, "cars")),
  cars = list(c("Merc 240D", "Fiat X1-9", "Ferrari Dino"))
)

df %>%
  mutate(filtered = map2(df, cars, ~ filter(.x, cars %in% .y)))
#> # A tibble: 1 × 3
#>   df             cars      filtered     
#>   <list>         <list>    <list>       
#> 1 <df [32 × 12]> <chr [3]> <df [3 × 12]>

df %>%
  mutate(filtered = map2(df, cars, ~ filter(.x, across(contains("cars"), ~ . %in% .y))))
#> Error in `mutate()`:
#> ! Problem while computing `filtered = map2(df, cars, ~filter(.x,
#>   across(contains("cars"), ~. %in% .y)))`.
#> Caused by error in `filter()`:
#> ! Problem while computing `..1 = across(contains("cars"), ~. %in% .y)`.
#> Caused by error in `across()`:
#> ! Problem while computing column `cars`.
#> Caused by error in `. %in% .y`:
#> ! the ... list contains fewer than 2 elements
```

<sup>Created on 2022-04-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

This is a known limitation of the way that `across()` currently works; we need to inline the generated calls for performance and sometimes our detection goes wrong. Unfortunately your use case is sufficiently arcane that we're unlikely to fix this problem specifically, but we are aware of the general problem and will consider it if we ever redesign `across()`.
